hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/AsyncJavaMigration.java

```diff
@@ -0,0 +1,151 @@
+abstract class AsyncJavaMigration<T> extends MirrorBaseJavaMigration {
+
+    private static final String CHECK_FLYWAY_SCHEMA_HISTORY_EXISTENCE_SQL = "select exists( " +
+            "  select * from information_schema.tables " +
+            "  where table_schema = :schema and table_name = 'flyway_schema_history')";
+
+    private static final String SELECT_LAST_CHECKSUM_SQL = "select checksum from flyway_schema_history " +
+            "where script = :className order by installed_rank desc limit 1";
+
+    private static final String UPDATE_CHECKSUM_SQL = "with last as (" +
+            "  select installed_rank from flyway_schema_history" +
+            "  where script = :className order by installed_rank desc limit 1" +
+            ") " +
+            "update flyway_schema_history f " +
+            "set checksum = :checksum " +
+            "from last " +
+            "where f.installed_rank = last.installed_rank";
+
+    protected final NamedParameterJdbcTemplate jdbcTemplate;
+
+    private final String schema;
+
+    private final TransactionOperations transactionOperations;
+
+    @Override
+    public Integer getChecksum() {
+        if (!hasFlywaySchemaHistoryTable()) {
+            return -1;
+        }
+
+        Integer lastChecksum = queryForObjectOrNull(SELECT_LAST_CHECKSUM_SQL, getSqlParamSource(), Integer.class);
+        if (lastChecksum == null) {
+            return -1;
+        } else if (lastChecksum < 0) {
+            return lastChecksum - 1;
+        } else if (lastChecksum != getSuccessChecksum()) {
+            return -1;
+        }
+        return lastChecksum;
+    }
+
+    @Override
+    public MigrationVersion getVersion() {
+        return null; // repeatable
+    }
+
+    public <O> O queryForObjectOrNull(String sql, SqlParameterSource paramSource, Class<O> requiredType) {
+        try {
+            return jdbcTemplate.queryForObject(sql, paramSource, requiredType);
+        } catch (EmptyResultDataAccessException ex) {
+            return null;
+        }
+    }
+
+    @Override
+    protected void doMigrate() throws IOException {
+        int checksum = getSuccessChecksum();
+        if (checksum <= 0) {
+            throw new IllegalArgumentException(String.format("Invalid non-positive success checksum %d", checksum));
+        }
+
+        Mono.fromRunnable(this::migrateAsync)
+                .subscribeOn(Schedulers.single())
+                .doOnSuccess(t -> onSuccess())
+                .doOnError(t -> log.error("Asynchronous migration failed:", t))
+                .subscribe();
+    }
+
+    protected void migrateAsync() {
+        long count = 0;
+        var last = Optional.of(getInitial());
+        var stopwatch = Stopwatch.createStarted();
+
+        try {
+            do {
+                final var previous = last;
+                last = transactionOperations.execute(t -> migratePartial(previous.get()));
+                count++;
+            } while (last != null && last.isPresent());
```

Comment:
`Optional` should not be null or it defeats the purpose. Can throw a `@Nonnull` on `migratePartial` to enforce.

```suggestion
            } while (last.isPresent());
```

Reply:
added `last != null` just to make sonar not to report a bug. Adding @Nonnull to `migratePartial` doesn't seem to make either IDEA or sonar to realize the return value can't be null.

Will make the change though.